### ISSUE TITLE
Data structure for glyphs

### ISSUE THREAD

**Issue by madig** [daltonmaag/Fontgardener#4]
Currently, glyph data is organized into sets into sources into layers into glyphs, sort of extending the UFO data model of putting glyphs in layers. 

1. It works, but introduces some inelegance when e.g. you want to export glyphs and want to determine, per layer, which other glyphs you have to also export because they're used in composites.
2. It leads to a lot of files on disk.

I wonder if there could be a more elegant data structure where e.g. you keep the sources with all glyphs and have a separate glyph->set mapping in a regular HashMap<NameOfGlyph, NameOfSet> :thinking_face:

Or even turn the relation on its head and have glyphs contain set name, sources and layers, sort of like Glyphs.app and FontLab 7 do.

You could even keep the structure on disk if desired and do the head-turning internally.

**Comment by belluzj** [comment]
Wasn't the point of separating glyphs into many "small" folders to help with dealing with changes in Git? If you bring everything back into one folder then you lose that

**Comment by madig** [comment]
Yes. That's why I thought it may be ok to leave the format on disk as is. Or you do a compromise where a glyph file contains all layers but you still have a glyph file per source. Real-world profiling would be needed to determine whether that would up the conflict rate or not...

**Comment by madig** [comment]
Idea: if we use the glyphs contains layers approach, we can have a top-level mapping of glyph names to glyphs, where glyphs have a "member_of_set: Option<Identifier>" field that says which set a glyph goes into. (Also a `unicodes: Vec<char>` field.) This would

1. flatten the fontgardener internal data structure while still allowing us to write every piece to a subfolder and its own file
    a. No need for extra glyph uniqueness checks
2. make set reassignment and splitting/merging sets as easy as changing the strings on a field
3. the string fields are similiar to "labels", which FontLab 7 (and Glyphs.app?)provides. This might make interop easier.

Component references would need some thinking then; Glyphs.app and FL7 have components reference the foreground/master layer.

**Comment by madig** [comment]
If we use the way Glyphs.app organizes layers in two levels, we could structure glyphs on disk like this:

```
- a/
--- Regular.json (contains default/foreground layer and the background and other misc. layers)
--- Bold.json (dito)
--- Italic.json (dito)
--- BoldItalic.json (dito)
```

I.e. instead of splitting everything into files, keep layers associated with a master in one file, because you typically change them together. This would be a compromise between too many files for a speedy workflow and to few to avoid conflicts.
